File: oxidize-python/oxidize_python/core/model/lora.py
"""LoRA adapters mirroring oxidize-golang/core/model/lora.go."""

from __future__ import annotations

from dataclasses import dataclass, field

from oxidize_python.core.model.model import Logits, Model, Session, Token


@dataclass
class LoraLayer:
    name: str
    rank: int
    alpha: float
    scale: float
    base_shape: list[int]
    up_loaded: bool = False
    down_loaded: bool = False
    up: list[float] = field(default_factory=list)
    down: list[float] = field(default_factory=list)
    in_dim: int = 0
    out_dim: int = 0

    def set_low_rank_weights(
        self, up: list[float], down: list[float], in_dim: int, out_dim: int
    ) -> None:
        self.up = up
        self.down = down
        self.in_dim = in_dim
        self.out_dim = out_dim
        self.up_loaded = len(up) > 0
        self.down_loaded = len(down) > 0

    def apply_low_rank_delta(self, x: list[float], out: list[float]) -> None:
        if not self.up_loaded or not self.down_loaded or self.rank <= 0:
            return
        if self.in_dim <= 0 or self.out_dim <= 0:
            return
        if len(x) < self.in_dim or len(out) < self.out_dim:
            return
        hidden = [0.0] * self.rank
        for r in range(self.rank):
            base = r * self.in_dim
            hidden[r] = sum(self.up[base + i] * x[i] for i in range(self.in_dim))
        scale = self.scale
        if scale == 0 and self.alpha > 0 and self.rank > 0:
            scale = self.alpha / self.rank
        for o in range(self.out_dim):
            delta = sum(self.down[o * self.rank + r] * hidden[r] for r in range(self.rank))
            out[o] += scale * delta


def new_lora_layer(name: str, rank: int, alpha: float, base_shape: list[int]) -> LoraLayer:
    scale = 1.0
    if alpha > 0 and rank > 0:
        scale = alpha / rank
    return LoraLayer(
        name=name,
        rank=rank,
        alpha=alpha,
        scale=scale,
        base_shape=base_shape,
    )


class LoraAdapter:
    def __init__(self, name: str, path: str, base: Model | None) -> None:
        self.name = name
        self.path = path
        self.layers: dict[str, LoraLayer] = {}
        self.rank = 0
        self.alpha = 0.0
        self.base_model = base

    def add_layer(self, layer: LoraLayer) -> None:
        self.layers[layer.name] = layer
        if layer.rank > 0 and self.rank == 0:
            self.rank = layer.rank
        if layer.alpha > 0 and self.alpha == 0:
            self.alpha = layer.alpha

    def apply(self, tokens: list[Token], session: Session) -> Logits:
        if self.base_model is None:
            raise LoraError("base model is nil")
        logits = self.base_model.forward(tokens, session)
        if self.rank > 0 and self.alpha > 0:
            scale = self.alpha / self.rank
            for i in range(len(logits)):
                logits[i] *= scale
        return logits

    def forward(self, tokens: list[Token], session: Session) -> Logits:
        return self.apply(tokens, session)

    def vocab_size(self) -> int:
        return self.base_model.vocab_size() if self.base_model else 0

    def context_size(self) -> int:
        return self.base_model.context_size() if self.base_model else 0

    def layer_count(self) -> int:
        return self.base_model.layer_count() if self.base_model else 0


class LoraError(Exception):
    def __init__(self, message: str) -> None:
        super().__init__(f"lora: {message}")


@dataclass
class LoraPlan:
    adapters: list[LoraAdapter] = field(default_factory=list)
    merge_strategy: str = "sequential"
    rank_budget: int = 0
    estimated_gain: float = 0.0

    def add(self, adapter: LoraAdapter) -> None:
        self.adapters.append(adapter)

    def validate(self) -> None:
        if not self.merge_strategy:
            raise LoraError("merge strategy is required")
        if not self.adapters:
            raise LoraError("no adapters in plan")
        if self.merge_strategy not in ("sequential", "parallel", "tying"):
            raise LoraError(f"unknown merge strategy: {self.merge_strategy}")

    def merge_adapters(self) -> LoraAdapter:
        merged = LoraAdapter("merged", "", None)
        merged.rank = self.rank_budget
        for a in self.adapters:
            for layer in a.layers.values():
                merged.add_layer(layer)
        merged.alpha = float(merged.rank) * self.estimated_gain
        return merged

    def estimate_memory(self) -> int:
        total = 0
        for a in self.adapters:
            for layer in a.layers.values():
                r = layer.rank or 1
                dim = 1
                for d in layer.base_shape:
                    dim *= d
                total += r * dim * 4
        return total

    def validate_rank_budget(self) -> None:
        for a in self.adapters:
            if a.rank > self.rank_budget:
                raise RankBudgetExceededError(a.rank, self.rank_budget)


class RankBudgetExceededError(Exception):
    def __init__(self, requested: int, available: int) -> None:
        super().__init__("lora: rank budget exceeded")
        self.requested = requested
        self.available = available


@dataclass
class LoraScalingConfig:
    auto: bool = False
    manual: float = 0.0
    max_gain: float = 0.0

    def scale_factor(self, rank: int) -> float:
        if self.auto:
            if rank == 0:
                return 0.0
            return min(self.max_gain, 1.0)
        return self.manual

File: oxidize-golang/core/model/lora.go
package model

import (
	"errors"
	"math"
)

// LoraLayer mirrors LoraLayer with optional low-rank weight matrices.
type LoraLayer struct {
	Name       string
	Rank       int
	Alpha      float32
	Scale      float32
	BaseShape  []int
	UpLoaded   bool
	DownLoaded bool
	Up         []float32 // [rank * inDim]
	Down       []float32 // [outDim * rank]
	InDim      int
	OutDim     int
}

// SetLowRankWeights attaches A/B matrices for low-rank adaptation.
func (l *LoraLayer) SetLowRankWeights(up, down []float32, inDim, outDim int) {
	l.Up, l.Down = up, down
	l.InDim, l.OutDim = inDim, outDim
	l.UpLoaded = len(up) > 0
	l.DownLoaded = len(down) > 0
}

// ApplyLowRankDelta adds scale * (x @ A @ B) to out when matrices are loaded.
func (l LoraLayer) ApplyLowRankDelta(x, out []float32) {
	if !l.UpLoaded || !l.DownLoaded || l.Rank <= 0 || l.InDim <= 0 || l.OutDim <= 0 {
		return
	}
	if len(x) < l.InDim || len(out) < l.OutDim {
		return
	}
	hidden := make([]float32, l.Rank)
	for r := 0; r < l.Rank; r++ {
		var sum float32
		base := r * l.InDim
		for i := 0; i < l.InDim; i++ {
			sum += l.Up[base+i] * x[i]
		}
		hidden[r] = sum
	}
	scale := l.Scale
	if scale == 0 && l.Alpha > 0 && l.Rank > 0 {
		scale = l.Alpha / float32(l.Rank)
	}
	for o := 0; o < l.OutDim; o++ {
		var sum float32
		for r := 0; r < l.Rank; r++ {
			sum += l.Down[o*l.Rank+r] * hidden[r]
		}
		out[o] += scale * sum
	}
}
// NewLoraLayer constructs a layer placeholder.
func NewLoraLayer(name string, rank int, alpha float32, baseShape []int) LoraLayer {
	scale := float32(1.0)
	if alpha > 0 && rank > 0 {
		scale = alpha / float32(rank)
	}
	return LoraLayer{Name: name, Rank: rank, Alpha: alpha, Scale: scale, BaseShape: baseShape}
}

// LoraAdapter mirrors LoraAdapter.
type LoraAdapter struct {
	Name     string
	Path     string
	Layers   map[string]LoraLayer
	Rank     int
	Alpha    float32
	BaseModel Model
}

// NewLoraAdapter constructs an empty adapter for a given base model.
func NewLoraAdapter(name, path string, base Model) *LoraAdapter {
	return &LoraAdapter{Name: name, Path: path, Layers: map[string]LoraLayer{}, BaseModel: base}
}

// AddLayer registers a LoRA layer with the adapter.
func (a *LoraAdapter) AddLayer(layer LoraLayer) {
	a.Layers[layer.Name] = layer
	if layer.Rank > 0 && a.Rank == 0 {
		a.Rank = layer.Rank
	}
	if layer.Alpha > 0 && a.Alpha == 0 {
		a.Alpha = layer.Alpha
	}
}

// Apply runs the adapter on top of the base model's forward pass.
func (a *LoraAdapter) Apply(tokens []Token, session *Session) (Logits, error) {
	if a.BaseModel == nil {
		return nil, errors.New("lora: base model is nil")
	}
	logits, err := a.BaseModel.Forward(tokens, session)
	if err != nil {
		return nil, err
	}
	if a.Rank > 0 && a.Alpha > 0 {
		scale := a.Alpha / float32(a.Rank)
		for i := range logits {
			logits[i] *= scale
		}
	}
	return logits, nil
}

// Forward implements Model by delegating to Apply.
func (a *LoraAdapter) Forward(tokens []Token, session *Session) (Logits, error) { return a.Apply(tokens, session) }

// LoraError mirrors LoraError.
type LoraError struct{ Message string }

func (e *LoraError) Error() string { return "lora: " + e.Message }

// LoraPlan mirrors LoraPlan.
type LoraPlan struct {
	Adapters      []*LoraAdapter
	MergeStrategy string // "sequential", "parallel", "tying"
	RankBudget    int
	EstimatedGain float32
}

// NewLoraPlan returns an empty plan.
func NewLoraPlan() *LoraPlan { return &LoraPlan{MergeStrategy: "sequential"} }

// Add attaches an adapter to the plan.
func (p *LoraPlan) Add(a *LoraAdapter) { p.Adapters = append(p.Adapters, a) }

// Validate returns nil if the plan is internally consistent.
func (p *LoraPlan) Validate() error {
	if p.MergeStrategy == "" {
		return &LoraError{Message: "merge strategy is required"}
	}
	if len(p.Adapters) == 0 {
		return &LoraError{Message: "no adapters in plan"}
	}
	switch p.MergeStrategy {
	case "sequential", "parallel", "tying":
	default:
		return &LoraError{Message: "unknown merge strategy: " + p.MergeStrategy}
	}
	return nil
}

// MergeAdapters produces a single consolidated LoraAdapter.
func (p *LoraPlan) MergeAdapters() *LoraAdapter {
	merged := NewLoraAdapter("merged", "", nil)
	merged.Rank = p.RankBudget
	for _, a := range p.Adapters {
		for _, l := range a.Layers {
			merged.AddLayer(l)
		}
	}
	merged.Alpha = float32(merged.Rank) * p.EstimatedGain
	return merged
}

// EstimateMemory estimates the memory cost in bytes for the plan.
func (p *LoraPlan) EstimateMemory() int64 {
	var total int64
	for _, a := range p.Adapters {
		for _, l := range a.Layers {
			r := int64(l.Rank)
			if r == 0 {
				r = 1
			}
			dim := 1
			for _, d := range l.BaseShape {
				dim *= d
			}
			total += int64(r) * int64(dim) * 4 // 2 fp32 matrices
		}
	}
	return total
}

// RankBudgetExceededError mirrors RankBudgetExceededError.
type RankBudgetExceededError struct{ Requested, Available int }

func (e *RankBudgetExceededError) Error() string {
	return "lora: rank budget exceeded"
}

// ValidateRankBudget ensures the requested rank does not exceed the budget.
func (p *LoraPlan) ValidateRankBudget() error {
	for _, a := range p.Adapters {
		if a.Rank > p.RankBudget {
			return &RankBudgetExceededError{Requested: a.Rank, Available: p.RankBudget}
		}
	}
	return nil
}

// LoraScalingConfig mirrors LoraScalingConfig.
type LoraScalingConfig struct {
	Auto    bool
	Manual  float32
	MaxGain float32
}

// ScaleFactor returns the LoRA scaling factor.
func (c LoraScalingConfig) ScaleFactor(rank int) float32 {
	if c.Auto {
		if rank == 0 {
			return 0
		}
		return float32(math.Min(float64(c.MaxGain), 1.0))
	}
	return c.Manual
}
